pkg/backend/telemetryapi/harvester.go

```diff
@@ -0,0 +1,535 @@
+	}
+	return reqs
+}
+
+func (h *Harvester) swapOutBatchMetrics() (req []request) {
+	h.lock.Lock()
+	rawMetricsBatch := h.rawMetricsBatch
+	h.rawMetricsBatch = nil
+	h.lock.Unlock()
+	var err error
+	req, err = newBatchRequest(rawMetricsBatch, h.config.APIKey, h.config.metricURL(), h.config.userAgent())
+	if err != nil {
+		h.config.logError(map[string]interface{}{
+			"err":     err.Error(),
+			"message": "error creating requests for batch metric",
+		})
+	}
+	return req
+}
+
+func (h *Harvester) swapOutSpans() []request {
+	h.lock.Lock()
+	sps := h.spans
+	h.spans = nil
+	h.lock.Unlock()
+
+	if nil == sps {
+		return nil
+	}
+	batch := &spanBatch{
+		AttributesJSON: h.commonAttributesJSON,
+		Spans:          sps,
+	}
+	reqs, err := newRequests(batch, h.config.APIKey, h.config.spanURL(), h.config.userAgent())
+	if nil != err {
+		h.config.logError(map[string]interface{}{
+			"err":     err.Error(),
+			"message": "error creating requests for spans",
+		})
+		return nil
+	}
+	return reqs
+}
+
+func harvestRequest(req request, cfg *Config) {
+	var attempts int
+	for {
+		cfg.logDebug(map[string]interface{}{
+			"event":       "data post",
+			"url":         req.Request.URL.String(),
+			"body-length": req.compressedBodyLength,
+		})
+		// Check if the audit log is enabled to prevent unnecessarily
+		// copying UncompressedBody.
+		if cfg.auditLogEnabled() {
+			cfg.logAudit(map[string]interface{}{
+				"event":   "uncompressed request body",
+				"url":     req.Request.URL.String(),
+				"data":    jsonString(req.UncompressedBody),
+				"headers": req.Request.Header,
+			})
+		}
+
+		resp := postData(req.Request, cfg.Client)
+
+		if nil != resp.err {
+			cfg.logError(map[string]interface{}{
+				"err": resp.err.Error(),
+			})
+		} else {
+			cfg.logDebug(map[string]interface{}{
+				"event":  "data post response",
+				"status": resp.statusCode,
+				"body":   jsonOrString(resp.body),
+			})
+		}
+		retry, backoff := resp.needsRetry(cfg, attempts)
+		if !retry {
+			return
+		}
+
+		tmr := time.NewTimer(backoff)
+		select {
+		case <-tmr.C:
+			break
+		case <-req.Request.Context().Done():
+			tmr.Stop()
+			return
+		}
+		attempts++
+
+		// Reattach request body because the original one has already been read
+		// and closed.
+		req.Request.Body = ioutil.NopCloser(bytes.NewBuffer(req.compressedBody))
+	}
+}
+
+// HarvestNow sends metric and span data to New Relic.  This method blocks until
+// all data has been sent successfully or the Config.HarvestTimeout timeout has
+// elapsed. This method can be used with a zero Config.HarvestPeriod value to
+// control exactly when data is sent to New Relic servers.
+func (h *Harvester) HarvestNow(ct context.Context) {
+	if nil == h {
+		return
+	}
+
+	ctx, cancel := context.WithTimeout(ct, h.config.HarvestTimeout)
+	defer cancel()
+
+	var reqs []request
+	reqs = append(reqs, h.swapOutMetrics(time.Now())...)
+	reqs = append(reqs, h.swapOutSpans()...)
+	reqs = append(reqs, h.swapOutBatchMetrics()...)
```

Comment:
Here we carry on as normal except added `swapOutBatchMetrics` to create the requests for the batch metrics.